File: app/src/main/java/com/stevenfrew/beatprompter/EventHandler.java
package com.stevenfrew.beatprompter;

import android.os.Handler;

import com.stevenfrew.beatprompter.pref.SettingsEventHandler;

public abstract class EventHandler extends Handler {
    private static final int HANDLER_MESSAGE_BASE_ID=1834739585;

    public static final int BLUETOOTH_MESSAGE_RECEIVED=HANDLER_MESSAGE_BASE_ID;
    public static final int CLIENT_CONNECTED=HANDLER_MESSAGE_BASE_ID+1;
    public static final int SERVER_CONNECTED=HANDLER_MESSAGE_BASE_ID+2;
    public static final int MIDI_START_SONG=HANDLER_MESSAGE_BASE_ID+3;
    public static final int MIDI_CONTINUE_SONG=HANDLER_MESSAGE_BASE_ID+4;
    public static final int MIDI_STOP_SONG=HANDLER_MESSAGE_BASE_ID+5;
    public static final int END_SONG=HANDLER_MESSAGE_BASE_ID+6;
    public static final int CLOUD_SYNC_ERROR=HANDLER_MESSAGE_BASE_ID+7;
    public static final int FOLDER_CONTENTS_FETCHING=HANDLER_MESSAGE_BASE_ID+8;
    public static final int MIDI_MSB_BANK_SELECT=HANDLER_MESSAGE_BASE_ID+9;
    public static final int MIDI_LSB_BANK_SELECT=HANDLER_MESSAGE_BASE_ID+10;
    public static final int MIDI_SONG_SELECT=HANDLER_MESSAGE_BASE_ID+11;
    public static final int MIDI_PROGRAM_CHANGE=HANDLER_MESSAGE_BASE_ID+12;
    public static final int SONG_LOAD_CANCELLED=HANDLER_MESSAGE_BASE_ID+13;
    public static final int SONG_LOAD_FAILED=HANDLER_MESSAGE_BASE_ID+14;
    public static final int SONG_LOAD_LINE_READ=HANDLER_MESSAGE_BASE_ID+15;
    public static final int SONG_LOAD_LINE_PROCESSED=HANDLER_MESSAGE_BASE_ID+16;
    public static final int SONG_LOAD_COMPLETED=HANDLER_MESSAGE_BASE_ID+17;
    public static final int MIDI_SET_SONG_POSITION=HANDLER_MESSAGE_BASE_ID+18;
    public static final int CACHE_UPDATED=HANDLER_MESSAGE_BASE_ID+19;
    public static final int SET_CLOUD_PATH=HANDLER_MESSAGE_BASE_ID+20;
    public static final int CLEAR_CACHE=HANDLER_MESSAGE_BASE_ID+21;
    public static final int FOLDER_CONTENTS_FETCHED=HANDLER_MESSAGE_BASE_ID+22;
    public static final int CLIENT_DISCONNECTED=HANDLER_MESSAGE_BASE_ID+23;
    public static final int SERVER_DISCONNECTED=HANDLER_MESSAGE_BASE_ID+24;

    private static final Object mSongListEventHandlerLock=new Object();
    private static final Object mSongDisplayEventHandlerLock=new Object();
    private static final Object mSettingsEventHandlerLock=new Object();
    private static SongListEventHandler mSongListEventHandler=null;
    private static SongDisplayEventHandler mSongDisplayEventHandler=null;
    private static Handler mSettingsEventHandler=null;

    public static void setSongListEventHandler(SongListEventHandler songListEventHandler)
    {
        synchronized(mSongListEventHandlerLock)
        {
            mSongListEventHandler = songListEventHandler;
        }
    }

    public static void setSongDisplayEventHandler(SongDisplayEventHandler songDisplayEventHandler)
    {
        synchronized(mSongDisplayEventHandlerLock) {
            mSongDisplayEventHandler = songDisplayEventHandler;
        }
    }

    public static void setSettingsEventHandler(SettingsEventHandler settingsEventHandler)
    {
        synchronized(mSettingsEventHandlerLock) {
            mSettingsEventHandler = settingsEventHandler;
        }
    }

    public static void sendEventToSongList(int event)
    {
        synchronized(mSongListEventHandlerLock) {
            if (mSongListEventHandler != null)
                mSongListEventHandler.obtainMessage(event).sendToTarget();
        }
    }

    public static void sendEventToSongDisplay(int event)
    {
        synchronized(mSongDisplayEventHandlerLock) {
            if (mSongDisplayEventHandler != null)
                mSongDisplayEventHandler.obtainMessage(event).sendToTarget();
        }
    }

    public static void sendEventToSettings(int event)
    {
        synchronized(mSettingsEventHandlerLock) {
            if (mSettingsEventHandler != null)
                mSettingsEventHandler.obtainMessage(event).sendToTarget();
        }
    }

    public static void sendEventToSongList(int event,Object arg)
    {
        synchronized(mSongListEventHandlerLock) {
            if (mSongListEventHandler != null)
                mSongListEventHandler.obtainMessage(event, arg).sendToTarget();
        }
    }

    public static void sendEventToSongDisplay(int event,Object arg)
    {
        synchronized(mSongDisplayEventHandlerLock) {
            if (mSongDisplayEventHandler != null)
                mSongDisplayEventHandler.obtainMessage(event, arg).sendToTarget();
        }
    }

    public static void sendEventToSettings(int event,Object arg)
    {
        synchronized(mSettingsEventHandlerLock) {
            if (mSettingsEventHandler != null)
                mSettingsEventHandler.obtainMessage(event, arg).sendToTarget();
        }
    }

    public static void sendEventToSongList(int event,int arg1,int arg2)
    {
        synchronized(mSongListEventHandlerLock) {
            if (mSongListEventHandler != null)
                mSongListEventHandler.obtainMessage(event, arg1, arg2).sendToTarget();
        }
    }

    public static void sendEventToSongDisplay(int event,int arg1,int arg2)
    {
        synchronized(mSongDisplayEventHandlerLock) {
            if (mSongDisplayEventHandler != null)
                mSongDisplayEventHandler.obtainMessage(event, arg1, arg2).sendToTarget();
        }
    }

    public static void sendEventToSettings(int event,int arg1,int arg2)
    {
        synchronized(mSettingsEventHandlerLock) {
            if (mSettingsEventHandler != null)
                mSettingsEventHandler.obtainMessage(event, arg1, arg2).sendToTarget();
        }
    }

    public static void sendEventToSongList(int event,int arg1,int arg2,Object arg3)
    {
        synchronized(mSongListEventHandlerLock) {
            if (mSongListEventHandler != null)
                mSongListEventHandler.obtainMessage(event, arg1, arg2, arg3).sendToTarget();
        }
    }

    public static void sendEventToSongDisplay(int event,int arg1,int arg2,Object arg3)
    {
        synchronized(mSongDisplayEventHandlerLock) {
            if (mSongDisplayEventHandler != null)
                mSongDisplayEventHandler.obtainMessage(event, arg1, arg2, arg3).sendToTarget();
        }
    }

    public static void sendEventToSettings(int event,int arg1,int arg2,Object arg3)
    {
        synchronized(mSettingsEventHandlerLock) {
            if (mSettingsEventHandler != null)
                mSettingsEventHandler.obtainMessage(event, arg1, arg2, arg3).sendToTarget();
        }
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/EventHandler.kt
package com.stevenfrew.beatprompter

import android.os.Handler
import com.stevenfrew.beatprompter.pref.SettingsFragment

abstract class EventHandler : Handler() {
    companion object {
        private const val HANDLER_MESSAGE_BASE_ID = 1834739585

        const val BLUETOOTH_CHOOSE_SONG = HANDLER_MESSAGE_BASE_ID
        const val CLIENT_CONNECTED = HANDLER_MESSAGE_BASE_ID + 1
        const val SERVER_CONNECTED = HANDLER_MESSAGE_BASE_ID + 2
        const val MIDI_START_SONG = HANDLER_MESSAGE_BASE_ID + 3
        const val MIDI_CONTINUE_SONG = HANDLER_MESSAGE_BASE_ID + 4
        const val MIDI_STOP_SONG = HANDLER_MESSAGE_BASE_ID + 5
        const val END_SONG = HANDLER_MESSAGE_BASE_ID + 6
        const val CLOUD_SYNC_ERROR = HANDLER_MESSAGE_BASE_ID + 7
        const val FOLDER_CONTENTS_FETCHING = HANDLER_MESSAGE_BASE_ID + 8
        const val FOLDER_CONTENTS_FETCHED = HANDLER_MESSAGE_BASE_ID + 9
        const val CLIENT_DISCONNECTED = HANDLER_MESSAGE_BASE_ID + 10
        const val MIDI_SONG_SELECT = HANDLER_MESSAGE_BASE_ID + 11
        const val MIDI_PROGRAM_CHANGE = HANDLER_MESSAGE_BASE_ID + 12
        const val SONG_LOAD_CANCELLED = HANDLER_MESSAGE_BASE_ID + 13
        const val SONG_LOAD_FAILED = HANDLER_MESSAGE_BASE_ID + 14
        const val SERVER_DISCONNECTED = HANDLER_MESSAGE_BASE_ID + 15
        const val SONG_LOAD_LINE_PROCESSED = HANDLER_MESSAGE_BASE_ID + 16
        const val SONG_LOAD_COMPLETED = HANDLER_MESSAGE_BASE_ID + 17
        const val MIDI_SET_SONG_POSITION = HANDLER_MESSAGE_BASE_ID + 18
        const val CACHE_UPDATED = HANDLER_MESSAGE_BASE_ID + 19
        const val SET_CLOUD_PATH = HANDLER_MESSAGE_BASE_ID + 20
        const val CLEAR_CACHE = HANDLER_MESSAGE_BASE_ID + 21
        const val BLUETOOTH_PAUSE_ON_SCROLL_START = HANDLER_MESSAGE_BASE_ID + 22
        const val BLUETOOTH_QUIT_SONG = HANDLER_MESSAGE_BASE_ID + 23
        const val BLUETOOTH_SET_SONG_TIME = HANDLER_MESSAGE_BASE_ID + 24
        const val BLUETOOTH_TOGGLE_START_STOP = HANDLER_MESSAGE_BASE_ID + 25

        private val mSongListEventHandlerLock = Any()
        private val mSongDisplayEventHandlerLock = Any()
        private val mSettingsEventHandlerLock = Any()
        private var mSongListEventHandler: SongList.SongListEventHandler? = null
        private var mSongDisplayEventHandler: SongDisplayActivity.SongDisplayEventHandler? = null
        private var mSettingsEventHandler: Handler? = null

        fun setSongListEventHandler(songListEventHandler: SongList.SongListEventHandler?) {
            synchronized(mSongListEventHandlerLock) {
                mSongListEventHandler = songListEventHandler
            }
        }

        fun setSongDisplayEventHandler(songDisplayEventHandler: SongDisplayActivity.SongDisplayEventHandler) {
            synchronized(mSongDisplayEventHandlerLock) {
                mSongDisplayEventHandler = songDisplayEventHandler
            }
        }

        fun setSettingsEventHandler(settingsEventHandler: SettingsFragment.SettingsEventHandler?) {
            synchronized(mSettingsEventHandlerLock) {
                mSettingsEventHandler = settingsEventHandler
            }
        }

        fun sendEventToSongList(event: Int) {
            synchronized(mSongListEventHandlerLock) {
                if (mSongListEventHandler != null)
                    mSongListEventHandler!!.obtainMessage(event).sendToTarget()
            }
        }

        fun sendEventToSongDisplay(event: Int) {
            synchronized(mSongDisplayEventHandlerLock) {
                if (mSongDisplayEventHandler != null)
                    mSongDisplayEventHandler!!.obtainMessage(event).sendToTarget()
            }
        }

        fun sendEventToSettings(event: Int) {
            synchronized(mSettingsEventHandlerLock) {
                if (mSettingsEventHandler != null)
                    mSettingsEventHandler!!.obtainMessage(event).sendToTarget()
            }
        }

        fun sendEventToSongList(event: Int, arg: Any) {
            synchronized(mSongListEventHandlerLock) {
                if (mSongListEventHandler != null)
                    mSongListEventHandler!!.obtainMessage(event, arg).sendToTarget()
            }
        }

        fun sendEventToSongDisplay(event: Int, arg: Any) {
            synchronized(mSongDisplayEventHandlerLock) {
                if (mSongDisplayEventHandler != null)
                    mSongDisplayEventHandler!!.obtainMessage(event, arg).sendToTarget()
            }
        }

        fun sendEventToSongDisplay(event: Int, arg1: Int, arg2: Int) {
            synchronized(mSongDisplayEventHandlerLock) {
                if (mSongDisplayEventHandler != null)
                    mSongDisplayEventHandler!!.obtainMessage(event, arg1, arg2).sendToTarget()
            }
        }
    }
}
